runtime/io.h

```diff
@@ -0,0 +1,207 @@
+// an error, following items will be ignored; but compiled code should
+// check for errors and avoid the following items when a READ statement
+// might have data dependences between items, as in
+//   READ(*,ERR=666) N, (A(J),J=1,N)
+// or when input/output items might raise a spurious fault, as in
+//   WRITE(*,ERR=666) X, CRASHIFCALLEDANDERRORPRESENT()
+bool IONAME(OutputDescriptor)(Cookie, const Descriptor &);
+bool IONAME(InputDescriptor)(Cookie, const Descriptor &);
+bool IONAME(OutputUnformattedBlock)(Cookie, const char *, std::size_t);
+bool IONAME(InputUnformattedBlock)(Cookie, char *, std::size_t);
+bool IONAME(OutputInteger64)(Cookie, std::int64_t);
+bool IONAME(InputInteger64)(Cookie, std::int64_t &, int kind = 8);
+bool IONAME(OutputReal32)(Cookie, float);
+bool IONAME(InputReal32)(Cookie, float &);
+bool IONAME(OutputReal64)(Cookie, double);
+bool IONAME(InputReal64)(Cookie, double &);
+bool IONAME(OutputComplex32)(Cookie, float, float);
+bool IONAME(OutputComplex64)(Cookie, double, double);
+bool IONAME(OutputAscii)(Cookie, const char *, std::size_t);
+bool IONAME(InputAscii)(Cookie, char *, std::size_t);
+bool IONAME(OutputLogical)(Cookie, bool);
+bool IONAME(InputLogical)(Cookie);
+
+std::size_t IONAME(GetSize)(Cookie);  // SIZE=
+
+// GetIoMsg() does not modify its argument unless an error or
+// end-of-record/file condition is present.
+void IONAME(GetIoMsg)(Cookie, char *, std::size_t);  // IOMSG=
+
+// The "errorHandling" argument to EndIoStatement() below conveys the
+// following flags that indicate how much error handling the
+// compiled code is able to perform.  If an error condition has arisen
+// that will not be handled, the runtime library will terminate the image
+// with a message.  Note that the runtime will never terminate the image
+// when (errorHandling & HasIostat) is true or when
+// (errorHanding == (HasErr|HasEof|HasEor)).
+enum ErrorHandlers {
+  NoErrorHandler = 0,
+  HasIostat = 1,  // IOSTAT= present
+  HasErr = 2,  // ERR= present
+  HasEof = 4,  // EOF= present
+  HasEor = 8,  // EOR= present
+};
+
+// This function must be called to end an I/O statement, and its
+// cookie value must dead to you afterwards (although it may be
+// recycled by the library and returned to serve a later I/O call).
+// The return value can be used to implement IOSTAT=, ERR=, END=, & EOR=;
+// store it into the IOSTAT= variable if there is one, and test
+// it to implement the various branches.
+int IONAME(EndIoStatement)(Cookie, int errorHandling = NoErrorHandler,
+    const char *sourceFileName = nullptr, int lineNumber = 0);
+
+// The value of IOSTAT= is zero when no error, end-of-record,
+// or end-of-file condition has arisen; errors are positive values.
+// (See 12.11.5 in Fortran 2018 for the complete requirements; some
+// of these constants must match the values of their corresponding
+// named constants in predefined modules.)
+enum Iostat {
```

Comment:
Eventually, some of these values must come from ISO_FORTRAN_ENV or possibly elsewhere.  It should be ok to defer setting that up for now.

Reply:
The information flow is probably in the other direction.  I'll set up a shared header suitable for use by both C and Fortran (although the copyright comment is going to be hard to do.)